File: sdks/python/nexus_sdk/transaction.py
"""Transaction support for Nexus SDK."""

from typing import Any, Dict, Optional
from enum import Enum

from nexus_sdk.client import NexusClient
from nexus_sdk.error import ValidationError
from nexus_sdk.models import QueryResult, Value


class TransactionStatus(Enum):
    """Transaction status."""

    ACTIVE = "active"
    COMMITTED = "committed"
    ROLLED_BACK = "rolled_back"
    NOT_STARTED = "not_started"


class Transaction:
    """Transaction handle for managing database transactions."""

    def __init__(self, client: NexusClient):
        """Create a new transaction handle.

        Args:
            client: NexusClient instance
        """
        self._client = client
        self._transaction_id: Optional[str] = None
        self._active: bool = False

    async def begin(self) -> None:
        """Begin a new transaction.

        Raises:
            ValidationError: If transaction is already active
            ApiError: If the API returns an error
        """
        if self._active:
            raise ValidationError("Transaction already active")

        await self._client.execute_cypher("BEGIN TRANSACTION", None)

        import asyncio

        self._active = True
        self._transaction_id = f"tx_{asyncio.get_event_loop().time()}"

    async def commit(self) -> None:
        """Commit the transaction.

        Raises:
            ValidationError: If no active transaction to commit
            ApiError: If the API returns an error
        """
        if not self._active:
            raise ValidationError("No active transaction to commit")

        await self._client.execute_cypher("COMMIT TRANSACTION", None)

        self._active = False
        self._transaction_id = None

    async def rollback(self) -> None:
        """Rollback the transaction.

        Raises:
            ValidationError: If no active transaction to rollback
            ApiError: If the API returns an error
        """
        if not self._active:
            raise ValidationError("No active transaction to rollback")

        await self._client.execute_cypher("ROLLBACK TRANSACTION", None)

        self._active = False
        self._transaction_id = None

    def is_active(self) -> bool:
        """Check if transaction is active.

        Returns:
            True if transaction is active, False otherwise
        """
        return self._active

    def status(self) -> TransactionStatus:
        """Get transaction status.

        Returns:
            TransactionStatus enum value
        """
        if self._active:
            return TransactionStatus.ACTIVE
        elif self._transaction_id is not None:
            return TransactionStatus.COMMITTED
        else:
            return TransactionStatus.NOT_STARTED

    async def execute(
        self, query: str, parameters: Optional[Dict[str, Value]] = None
    ) -> QueryResult:
        """Execute a Cypher query within this transaction.

        Args:
            query: Cypher query string
            parameters: Optional query parameters

        Returns:
            QueryResult containing query results

        Raises:
            ValidationError: If transaction is not active
            ApiError: If the API returns an error
        """
        if not self._active:
            raise ValidationError("Transaction is not active")

        return await self._client.execute_cypher(query, parameters)

    @property
    def transaction_id(self) -> Optional[str]:
        """Get the transaction ID.

        Returns:
            Transaction ID or None
        """
        return self._transaction_id

File: sdks/rust/src/transaction.rs
//! Transaction support for Nexus SDK

use crate::client::NexusClient;
use crate::error::{NexusError, Result};

/// Transaction handle for managing database transactions
#[derive(Debug, Clone)]
pub struct Transaction {
    client: NexusClient,
    transaction_id: Option<String>,
    active: bool,
}

/// Transaction status
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TransactionStatus {
    /// Transaction is active
    Active,
    /// Transaction has been committed
    Committed,
    /// Transaction has been rolled back
    RolledBack,
    /// Transaction is not started
    NotStarted,
}

impl Transaction {
    /// Create a new transaction handle
    pub(crate) fn new(client: NexusClient) -> Self {
        Self {
            client,
            transaction_id: None,
            active: false,
        }
    }

    /// Begin a new transaction
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use nexus_sdk::NexusClient;
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), nexus_sdk::NexusError> {
    /// # let client = NexusClient::new("http://localhost:15474")?;
    /// let mut tx = client.begin_transaction().await?;
    /// // Perform operations...
    /// tx.commit().await?;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn begin(&mut self) -> Result<()> {
        if self.active {
            return Err(NexusError::Validation(
                "Transaction already active".to_string(),
            ));
        }

        let _result = self
            .client
            .execute_cypher("BEGIN TRANSACTION", None)
            .await?;

        self.active = true;
        self.transaction_id = Some(format!(
            "tx_{}",
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_nanos()
        ));

        Ok(())
    }

    /// Commit the transaction
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use nexus_sdk::{NexusClient, Transaction};
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), nexus_sdk::NexusError> {
    /// # let client = NexusClient::new("http://localhost:15474")?;
    /// # let mut tx: Transaction = client.begin_transaction().await?;
    /// tx.commit().await?;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn commit(&mut self) -> Result<()> {
        if !self.active {
            return Err(NexusError::Validation(
                "No active transaction to commit".to_string(),
            ));
        }

        let _result = self
            .client
            .execute_cypher("COMMIT TRANSACTION", None)
            .await?;

        self.active = false;
        self.transaction_id = None;

        Ok(())
    }

    /// Rollback the transaction
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use nexus_sdk::{NexusClient, Transaction};
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), nexus_sdk::NexusError> {
    /// # let client = NexusClient::new("http://localhost:15474")?;
    /// # let mut tx: Transaction = client.begin_transaction().await?;
    /// tx.rollback().await?;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn rollback(&mut self) -> Result<()> {
        if !self.active {
            return Err(NexusError::Validation(
                "No active transaction to rollback".to_string(),
            ));
        }

        let _result = self
            .client
            .execute_cypher("ROLLBACK TRANSACTION", None)
            .await?;

        self.active = false;
        self.transaction_id = None;

        Ok(())
    }

    /// Check if transaction is active
    pub fn is_active(&self) -> bool {
        self.active
    }

    /// Get transaction status
    pub fn status(&self) -> TransactionStatus {
        if self.active {
            TransactionStatus::Active
        } else if self.transaction_id.is_some() {
            TransactionStatus::Committed
        } else {
            TransactionStatus::NotStarted
        }
    }

    /// Execute a Cypher query within this transaction
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use nexus_sdk::{NexusClient, Transaction};
    /// # use std::collections::HashMap;
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), nexus_sdk::NexusError> {
    /// # let client = NexusClient::new("http://localhost:15474")?;
    /// # let mut tx: Transaction = client.begin_transaction().await?;
    /// let result = tx.execute("CREATE (n:Person {name: 'Alice'}) RETURN n", None).await?;
    /// tx.commit().await?;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn execute(
        &self,
        query: &str,
        params: Option<std::collections::HashMap<String, crate::models::Value>>,
    ) -> Result<crate::models::QueryResult> {
        if !self.active {
            return Err(NexusError::Validation(
                "Transaction is not active".to_string(),
            ));
        }

        self.client.execute_cypher(query, params).await
    }
}

impl NexusClient {
    /// Begin a new transaction
    ///
    /// # Example
    ///
    /// ```no_run
    /// # use nexus_sdk::{NexusClient, Transaction};
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), nexus_sdk::NexusError> {
    /// # let client = NexusClient::new("http://localhost:15474")?;
    /// let mut tx: Transaction = client.begin_transaction().await?;
    /// // Perform operations...
    /// tx.commit().await?;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn begin_transaction(&self) -> Result<Transaction> {
        let mut tx = Transaction::new(self.clone());
        tx.begin().await?;
        Ok(tx)
    }
}
